### PR TITLE
Azure

### PR DESCRIPTION
**Pull request by ivoflipse**
Recipes for some of our azure package. Preferably I'd add all important ones, but not sure whether it's preferred to have them all in separate repos or not

### CODE REVIEW

**Comment by ivoflipse** [comment]
Sadly for me, turns out that this version won't be compatible with azure-storage version 0.32.0, so I'll have to do this whole mess over again...

**Comment by msarahan** [comment]
Nice, clean recipes.  Are there test suites for any of these?  Import tests are OK, but running test suites is a much better guarantee of good packages.

Do you want to hold off merging until the azure-storage comes out, or do you want this available sooner?

**Comment by ivoflipse** [comment]
I think I'll continue to mess around with it for now, though it's too bad I can't download the packages that the CI produces. It seems there are some useful packages in here after all, since the new azure-storage package only requires azure-common (which is in here).

The problem for me seemed to be that the azure package depends on an older version of azure-storage, but luckily it's not vice versa.

I'll look into getting their tests working, because they do have them.

File: recipes/azure-common/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,40 @@
+{% set version = "1.1.3" %}
+
+package:
+  name: azure-common
+  version: {{ version }}
+
+source:
+  fn: azure-common-{{ version }}.zip
+  url: https://pypi.python.org/packages/48/89/017df9bed55925de004076937ca168d3a03aa0250fbda7826ec3c0ac714f/azure-common-{{ version }}.zip
```
Comment (jakirkham) [inline recipes/azure-common/meta.yaml]: Could we please change this to `https://pypi.io/packages/source/a/azure-common/azure-common-{{ version }}.zip`?

File: recipes/azure-mgmt-common/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,45 @@
+{% set version = "0.20.0" %}
+
+package:
+  name: azure-mgmt-common
+  version: {{ version }}
+
+source:
+  fn: azure-mgmt-common{{ version }}.zip
+  url: https://pypi.python.org/packages/source/a/azure-mgmt-common/azure-mgmt-common-{{ version }}.zip
```
Comment (jakirkham) [inline recipes/azure-mgmt-common/meta.yaml]: Please change `pypi.python.org` -> `pypi.io`.

File: recipes/azure-mgmt-compute/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,41 @@
+{% set version = "0.20.1" %}
+
+package:
+  name: azure-mgmt-compute
+  version: {{ version }}
+
+source:
+  fn: azure-mgmt-compute-{{ version }}.zip
+  url: https://pypi.python.org/packages/source/a/azure-mgmt-compute/azure-mgmt-compute-{{ version }}.zip
```
Comment (jakirkham) [inline recipes/azure-mgmt-compute/meta.yaml]: Please change `pypi.python.org` -> `pypi.io`.

File: recipes/azure-mgmt-network/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,42 @@
+{% set version = "0.20.1" %}
+
+package:
+  name: azure-mgmt-network
+  version: {{ version }}
+
+source:
+  fn: azure-mgmt-network-{{ version }}.zip
+  url: https://pypi.python.org/packages/source/a/azure-mgmt-network/azure-mgmt-network-{{ version }}.zip
```
Comment (jakirkham) [inline recipes/azure-mgmt-network/meta.yaml]: Please change `pypi.python.org` -> `pypi.io`.

File: recipes/azure-mgmt-nspkg/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,41 @@
+{% set version = "1.0.0" %}
+
+package:
+  name: azure-mgmt-nspkg
+  version: {{ version }}
+
+source:
+  fn: azure-mgmt-nspkg-{{ version }}.zip
+  url: https://pypi.python.org/packages/source/a/azure-mgmt-nspkg/azure-mgmt-nspkg-{{ version }}.zip
```
Comment (jakirkham) [inline recipes/azure-mgmt-nspkg/meta.yaml]: Please change `pypi.python.org` -> `pypi.io`.

File: recipes/azure-mgmt-resource/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,41 @@
+{% set version = "0.20.1" %}
+
+package:
+  name: azure-mgmt-resource
+  version: {{ version }}
+
+source:
+  fn: azure-resource-{{ version }}.zip
+  url: https://pypi.python.org/packages/source/a/azure-mgmt-resource/azure-mgmt-resource-{{ version }}.zip
```
Comment (jakirkham) [inline recipes/azure-mgmt-resource/meta.yaml]: Please change `pypi.python.org` -> `pypi.io`.

File: recipes/azure-mgmt/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,50 @@
+{% set version = "0.20.1" %}
+
+package:
+  name: azure-mgmt
+  version: {{ version }}
+
+source:
+  fn: azure-mgmt-{{ version }}.zip
+  url: https://pypi.python.org/packages/source/a/azure-mgmt/azure-mgmt-{{ version }}.zip
```
Comment (jakirkham) [inline recipes/azure-mgmt/meta.yaml]: Please change `pypi.python.org` -> `pypi.io`.

File: recipes/azure-mgmt-storage/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,41 @@
+{% set version = "0.20.0" %}
+
+package:
+  name: azure-mgmt-storage
+  version: {{ version }}
+
+source:
+  fn: azure-mgmt-storage-{{ version }}.zip
+  url: https://pypi.python.org/packages/source/a/azure-mgmt-storage/azure-mgmt-storage-{{ version }}.zip
```
Comment (jakirkham) [inline recipes/azure-mgmt-storage/meta.yaml]: Please change `pypi.python.org` -> `pypi.io`.

File: recipes/azure-nspkg/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,38 @@
+{% set version = "1.0.0" %}
+
+package:
+  name: azure-nspkg
+  version: {{ version }}
+
+source:
+  fn: azure-nspkg-{{ version }}.zip
+  url: https://pypi.python.org/packages/source/a/azure-nspkg/azure-nspkg-{{ version }}.zip
```
Comment (jakirkham) [inline recipes/azure-nspkg/meta.yaml]: Please change `pypi.python.org` -> `pypi.io`.

File: recipes/azure-servicebus/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,44 @@
+{% set version = "0.20.1" %}
+
+package:
+  name: azure-servicebus
+  version: {{ version }}
+
+source:
+  fn: azure-servicebus-{{ version }}.zip
+  url: https://pypi.python.org/packages/source/a/azure-servicebus/azure-servicebus-{{ version }}.zip
```
Comment (jakirkham) [inline recipes/azure-servicebus/meta.yaml]: Please change `pypi.python.org` -> `pypi.io`.

File: recipes/azure-servicemanagement-legacy/meta.yaml [hunk 11]
```diff
@@ -0,0 +1,46 @@
+{% set version = "0.20.1" %}
+
+package:
+  name: azure-servicemanagement-legacy
+  version: {{ version }}
+
+source:
+  fn: azure-servicemanagement-legacy-{{ version }}.zip
+  url: https://pypi.python.org/packages/5b/4b/013c5b8e782a7b744a6d92a5eb33e884615b9101bace5ddfd73c2e16fef3/azure-servicemanagement-legacy-{{ version }}.zip
```
Comment (jakirkham) [inline recipes/azure-servicemanagement-legacy/meta.yaml]: Please change to `https://pypi.io/packages/source/a/azure-servicemanagement-legacy/azure-servicemanagement-legacy-{{ version }}.zip`

File: recipes/azure/meta.yaml [hunk 12]
```diff
@@ -0,0 +1,46 @@
+{% set version = "1.0.3" %}
+
+package:
+  name: azure
+  version: {{ version }}
+
+source:
+  fn: azure-{{ version }}.zip
+  url: https://pypi.python.org/packages/source/a/azure/azure-{{ version }}.zip
```
Comment (jakirkham) [inline recipes/azure/meta.yaml]: Please change `pypi.python.org` -> `pypi.io`.

File: recipes/azure-storage/meta.yaml [hunk 13]
```diff
@@ -0,0 +1,53 @@
+{% set version = "0.20.3" %}
+
+package:
+  name: azure-storage
+  version: {{ version }}
+
+source:
+  fn: azure-storage-{{ version }}.zip
+  url: https://pypi.python.org/packages/9b/ce/60a1173299afc98814e582c419cf84b358c083ad86fb21bafd801548ca5b/azure-storage-{{ version }}.zip
```
Comment (jakirkham) [inline recipes/azure-storage/meta.yaml]: Please change to `https://pypi.io/packages/source/a/azure-storage/azure-storage-{{ version }}.zip`.

**Comment by jakirkham** [comment]
@Korijn, @Maxyme, could both of you please confirm that it is alright to be added as maintainers of these recipes?

File: recipes/azure-common/meta.yaml [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+    - azure-nspkg
+
+test:
+  imports:
+    - azure
+    - azure.common
+
+about:
+  home: https://github.com/Azure/azure-sdk-for-python
+  license: Apache License 2.0
```
Comment (jakirkham) [inline recipes/azure-common/meta.yaml]: Normally, the short form is preferred. In other words, dropping license.

File: recipes/azure-common/meta.yaml [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+  name: azure-common
+  version: {{ version }}
+
+source:
+  fn: azure-common-{{ version }}.zip
+  url: https://pypi.python.org/packages/48/89/017df9bed55925de004076937ca168d3a03aa0250fbda7826ec3c0ac714f/azure-common-{{ version }}.zip
+  md5: 13471213f49400c5d483865c96558ab9
+  
+build:
+  script: pip install --no-deps .
```
Comment (jakirkham) [inline recipes/azure-common/meta.yaml]: Normally, we use `python setup.py install --single-version-externally-managed --record=record.txt` for `setuptools`-based installs and `python setup.py install` for `distutils`-based install. Though there are a few places where `pip` ends up being used too.

**Comment by jakirkham** [comment]
Honestly, I'm not too worried about rigorous testing here as I believe these are all pure Python. While I'm generally a fan of rigorous testing with compiled code, it doesn't feel that doing this is strictly necessary with pure Python code. It is normally has tests, goes through CI, and doesn't suffer from the same problems compiled code does. As such it is generally in good shape at release time.

Basically, this looks ok to me. Would be nice to see some cosmetic changes, but none of them are so crucial.

**Comment by Korijn** [comment]
I'm actually working on these recipes right now, so yes, I'd love to be listed as a maintainer.

**Comment by Maxyme** [comment]
@jakirkham Yes, you can add me as maintainer of this recipe.

File: recipes/azure-common/meta.yaml (hunk 15 above)
Comment (ivoflipse) [inline recipes/azure-common/meta.yaml]: I took the pip install from the discussion, where it seemed people were happy with it. But I've switched it to python setup.py install instead

File: recipes/azure/meta.yaml [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+    - azure-mgmt ==0.20.1
+    - azure-servicebus ==0.20.1
+    - azure-storage ==0.20.3
+    - azure-servicemanagement-legacy ==0.20.1
+  run:
+    - python
+    - azure-mgmt ==0.20.1
+    - azure-servicebus ==0.20.1
+    - azure-storage ==0.20.3
+    - azure-servicemanagement-legacy ==0.20.1
```
Comment (Korijn) [inline recipes/azure/meta.yaml]: these pinned versions seem to be outdated (and are you sure they're not too restrictive?)

File: recipes/azure-storage/meta.yaml [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+    - requests
+    - futures
+    - vcrpy
+  run:
+    - python
+    - azure-common
+    - python-dateutil
+    - requests
+    - futures
+    - vcrpy
```
Comment (Korijn) [inline recipes/azure-storage/meta.yaml]: I think this is a testing requirement

File: recipes/azure-storage/meta.yaml [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+  run:
+    - python
+    - azure-common
+    - python-dateutil
+    - requests
+    - futures
```
Comment (Korijn) [inline recipes/azure-storage/meta.yaml]: Needs `# [py2k]`

File: recipes/azure-storage/meta.yaml [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+  number: 0
+
+requirements:
+  build:
+    - python
+    - setuptools
+    - azure-common
+    - python-dateutil
+    - requests
+    - futures
```
Comment (Korijn) [inline recipes/azure-storage/meta.yaml]: Needs `# [py2k]`

File: recipes/azure-storage/meta.yaml [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+    - vcrpy
```
Comment (Korijn) [inline recipes/azure-storage/meta.yaml]: testing requirement?

File: recipes/azure-mgmt/meta.yaml [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+    - azure-mgmt-compute ==0.20.1
+    - azure-mgmt-network ==0.20.1
+    - azure-mgmt-resource ==0.20.1
+    - azure-mgmt-storage ==0.20.0
+  run:
+    - python
+    - azure-mgmt-compute ==0.20.1
+    - azure-mgmt-network ==0.20.1
+    - azure-mgmt-resource ==0.20.1
+    - azure-mgmt-storage ==0.20.0
```
Comment (Korijn) [inline recipes/azure-mgmt/meta.yaml]: outdated versions?

File: recipes/azure/meta.yaml (hunk 16 above)
Comment (ivoflipse) [inline recipes/azure/meta.yaml]: Yes, they probably are. One downside though, is that figuring out all the dependencies isn't as straightforward, because they cut everything up into tiny pieces with disparate versions

Comment (Korijn) [inline recipes/azure/meta.yaml]: You only need to be as specific as the azure sdk for python developers are in their `setup.py`, no?

Comment (ivoflipse) [inline recipes/azure/meta.yaml]: https://github.com/Azure/azure-sdk-for-python/blob/v1.0.3/azure/setup.py Their setup.py is actually VERY restrictive

File: recipes/azure-storage/meta.yaml [hunk 22]
```diff
@@ -17,20 +17,22 @@ requirements:
...
     - azure-common
     - python-dateutil
     - requests
-    - futures
-    - vcrpy
+    - futures  # [py2k]
 
 test:
+  requires:
+    - - vcrpy
```
Comment (Korijn) [inline recipes/azure-storage/meta.yaml]: typo

File: recipes/azure-mgmt-storage/meta.yaml (hunk 8 above)
Comment (Korijn) [inline recipes/azure-mgmt-storage/meta.yaml]: You missed one @ivoflipse

**Comment by ivoflipse** [comment]
We'll first need to merge this, then bump azure-storage on the feedstock

File: recipes/azure-common/meta.yaml (hunk 15 above)
Comment (ocefpaf) [inline recipes/azure-common/meta.yaml]: > I took the pip install from the discussion, where it seemed people were
happy with it. But I've switched it to python setup.py install instead

That discussion never settled on a new standard and we ended up using the
"old way," but you are correct. We considered pip and we may be using it
again eventually.

File: recipes/azure-mgmt/meta.yaml (hunk 21 above)
Comment (ivoflipse) [inline recipes/azure-mgmt/meta.yaml]: Those are specifically the ones requires by the setup.py: https://github.com/Azure/azure-sdk-for-python/blob/v1.0.3/azure-mgmt/setup.py

All packages have to be updated in tandem for this to work :(

Comment (Korijn) [inline recipes/azure-mgmt/meta.yaml]: OK, then it makes sense to me.

**Comment by Korijn** [comment]
LGTM

File: recipes/azure-mgmt/meta.yaml (hunk 21 above)
Comment (jakirkham) [inline recipes/azure-mgmt/meta.yaml]: One thing you might consider to make life easier here is to just use `{{ version }}` for all of these as they seem to many seem to share the same version as the package itself. This might make updating a little nicer.

Also, I think ` =={{ version}}` is the same as ` {{ version}}`.

If there are a couple that always lag, we might be able to come up with some programmatic to describe that.

All of this is totally up to you and not required. Just thinking of ways to make updating this package easier.

File: recipes/azure-nspkg/meta.yaml [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+    - python
+    - setuptools
+  
+  run:
+    - python
+    - setuptools
+
+test:
+  requires:
+    - setuptools
```
Comment (jakirkham) [inline recipes/azure-nspkg/meta.yaml]: This should be handled by the `run` requirement already; so, can be removed.

File: recipes/azure-common/meta.yaml [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+test:
+  imports:
+    - azure
+    - azure.common
+  commands:
+    - python -m unittest discover -p "test_*.py"
+
+about:
+  home: https://github.com/Azure/azure-sdk-for-python
+  license: Apache License 2.0
```
Comment (jakirkham) [inline recipes/azure-common/meta.yaml]: This is a little tedious (sorry), but we tend to prefer shorter license strings. So, this would be better as `Apache 2.0`, `Apache-2.0`, or similar.

Comment (jakirkham) [inline recipes/azure-common/meta.yaml]: I won't comment on all of these. Suffice it to say this would be a welcome change for the other recipes too.

**Comment by jakirkham** [comment]
> @jakirkham I applied the suggested changes and added tests after all, which seemed to notice their pkg_resources shenanigans, so I guess it was useful after all

Yeah, we made a change to the build system recently. Normally, `conda` installs `setuptools`, `pip`, and `wheel` with `python` by default. We disabled this at staged-recipes (not yet disabled in feedstocks though planned) as it made things a bit confusing and caused various problems. This probably caused these issues. We have caught at least one surprising case like `azure-nspkg` where something required `setuptools` at run-time because of this change.

File: recipes/azure-servicebus/meta.yaml [hunk 25]
```diff
@@ -0,0 +1,45 @@
...
+  imports:
+    - azure
+    - azure.servicebus
+    - azure.servicebus._http
+  commands:
+    - python -m unittest discover -p "test_*.py"
+
+about:
+  home: https://github.com/Azure/azure-sdk-for-python
+  license: Apachage Software License
```
Comment (jakirkham) [inline recipes/azure-servicebus/meta.yaml]: Typo? Also, please make sure there is a version.

File: recipes/azure-mgmt/meta.yaml (hunk 21 above)
Comment (ivoflipse) [inline recipes/azure-mgmt/meta.yaml]: @jakirkham that's something I'll try to keep in mind. I reckon that indeed they have version numbers that somewhat increase in tandem, so this could save some work

**Comment by jakirkham** [comment]
LGTM. Any other thoughts on this?

**Comment by ivoflipse** [comment]
No @jakirkham I think its good to go

**Comment by jakirkham** [comment]
Thanks @ivoflipse.

**Comment by jakirkham** [comment]
Looks like AppVeyor is having some difficulty registering at some point. Restarting...

**Comment by Korijn** [comment]
I noticed the builds on the feedstocks are failing for most of these, since they started at the same time and are all interdependent. So some restarting of CI builds is probably in order...

**Comment by jakirkham** [comment]
Ack! GitHub rate-limit kicked in. This imposes a ~20min timeout. We will have to wait to restart the rest.

I may just clear out the ones that appear to be done so that we have fewer to go through in the next round.

**Comment by jakirkham** [comment]
I have removed the ones that appear to have completed the feedstock generation process.

**Comment by jakirkham** [comment]
I have temporarily removed the ones that did not clear feedstock generation with this PR ( https://github.com/conda-forge/staged-recipes/pull/798 ) and will re-add them with this PR ( https://github.com/conda-forge/staged-recipes/pull/799 ) when we are ready again.

**Comment by ivoflipse** [comment]
Sorry for all the trouble @jakirkham!

**Comment by jakirkham** [comment]
No trouble at all.

The GitHub API rate-limit is something we are well aware of. In fact, I was trying to use this as a way to test further optimizations of our use the GitHub API. Not sure that it was going to make it all the way through, but I wanted to see how far it would get. Looks like it went 9 recipes and may have completed the 10th if AppVeyor didn't time out during CI registration. While this doesn't sound amazing, we use to get throttled every one recipe. So, this is a pretty big improvement.

The next batch should go through smoothly (looks like only 4). Please let me know if anything appears off about the 9 that did go through. Would be happy to investigate.

**Comment by jakirkham** [comment]
These should now all be cleared. Thanks for your patience. Please let us know if there is anything missing or if there are any issues with these that needs correcting.

**Comment by ivoflipse** [comment]
Still not entirely cleared, probably because of the dependency chain. So everything except azure-nspkg  failed. Now azure-common has to be built, before some of the others will succeed

**Comment by jakirkham** [comment]
Sorry, by cleared I meant converted to a feedstock. Built is a different question. As far as Travis CI and CircleCI, you should have permissions on those. Though you may need to resync or something (particularly with Travis CI).

AppVeyor is a different story as we can't granularize permissions. If that is messed up, please tell us what order they need to be in and we should be able to fix that.
